### DISCUSSION TITLE
Adafruit Macropad / CircuitPython

### DISCUSSION THREAD

**Discussion by nomadicfrog** [rsjaffe/MIDI2LR#632] [Q&A]
Hi everyone,

I'm trying to wrap my head around the wide range of factors involved with getting this all to work, and I thought I'd reach out to see if somebody has already done what I'm attempting. I've searched this forum and Google more broadly, and haven't found anyone else. If I missed an existing thread, I apologize.

TL;DR: is anyone using MIDI2LR with this [Adafruit macropad kit](https://www.adafruit.com/product/5128)?

Here's what I have and where I'm at:

- I just bought and assembled an Adafruit [macropad kit](https://www.adafruit.com/product/5128). This DIY kit has 12 keys and a rotary encoder (which also clicks when pushed) so those are the input options I am working with. Most of the YouTube videos and other support I've found for MIDI2LR revolve around much larger off-the-shelf macropads, editing decks, etc.
- I've installed the MIDI2LR plugin.
- I've found some example CircuitPython code for the macropad that sends MIDI commands.
- I can get the macropad to interact with Lightroom, and anything that is a straightforward key command (such as "r" to crop) works.
- I am still trying to figure out more complex moves, though. More on that in a second.

If it matters, I'm using the latest Lightroom CC on macOS 11.6.1. And to be clear: I don't have a problem, everything seems to be working fine so far. I'm just checking to see if anyone has been here before me and has anything I can use as a starting point. If not, I'll try to report back when I figure it all out in case somebody like me comes looking.

My first goal is pretty simple, I think:

1. Press a key on the 3x4 pad to select the Develop attribute I want to adjust (for example, Temperature, Tint, Exposure, etc.)
2. Turn the rotary encoder to adjust that attribute.
3. Repeat 1 & 2 as necessary for each attribute.

To start with I thought I'd assign the following keys, with the rotary encoder dynamically changing to adjust the corresponding value.

Keys

1. Temp
2. Tint
3. Exposure
4. Contrast
5. Highlights
6. Shadows
7. Whites
8. Blacks
9. Clarity
10. Transform: Rotate
11. not sure yet
12. Toggle: Lens Corrections / Enable Profile Corrections

So, my understanding - correct me if I'm wrong - is that I need to re-write the CircuitPython code on the macropad to handle my internal logic: basically to change the MIDI code the rotary encoder sends to MIDI2LR to be distinct depending on if I am talking about Temperature, Tint, Exposure, etc.

In fact, I think I just realized, I'm not sure I even need to map my macropad keys to MIDI2LR at all...it's all about what MIDI codes the rotary encoder sends?

Anyhow I couldn't find anyone using this Adafruit [macropad](https://www.adafruit.com/product/5128), and I thought I'd see if I had missed it. Maybe there are existing profiles and / or CircuitPython code I could use as a starting point.

As this is all rather complicated I'm having a hard time figuring out which chicken / egg to work on first.

Thanks for any pointers! This looks like a great plugin and a good community!

-Mark

**Comment by nomadicfrog** [comment]
I think the first thing I'm trying to figure out is: what do I send to MIDI2LR? I'm trying to understand the difference between "control change", "note", and "pitch bend".

The example code I've found seems to be inconsistent from one source to another, and I can't find MIDI2LR documentation that addresses that more-raw end of the process.

https://learn.adafruit.com/adafruit-macropad-rp2040/macropad-midi

https://learn.adafruit.com/neocontroller-color-grading-input-box

Everything I find here and in YouTube setup tutorials seem to deal with pre-packaged devices and you just have to map whatever the device outputs to MIDI2LR, whereas I am in the position of needing to actually _determine_ what the device sends out, and I'm not sure what to have it "say".

**Comment by rsjaffe** [comment]
Best is control change. Also, for the encoder, I'd recommend using relative mode rather than absolute. This allows you to adjust the sensitivity. Read the entirety of https://github.com/rsjaffe/MIDI2LR/wiki/MIDI-Controller-Setup to get a good idea of what to do.

**Comment by fyrus7** [comment]
Did you find the answer yet? I made my own macropad with arduino pro micro and sent midi cc to midi2lr.

I use any midi cc value from 1 to anything (now up to 60) and set my encoder to sent cc 63 & 65 for value up and down.
